Oracle connector read all tables on ALL schemas under the same DB, not the only schema's table the user id can access

<!-- 
Welcome to Airbyte!
We're really appreciate your report and know that this will help us build an 
amazing tool. If you want to contribute yourself, you can find a good place 
to start by searching for the good-first-issues label or maybe... by trying 
to solve this one? (we can help debug this with you!)

Right now we are in alpha, so we're releasing versions a lot more frequently than 
normal. You can help us get to the root of the problem faster by filling out the
questionnaire below!

It's really important having all information and context. 
You can remove the examples bellow and fill out with your information.

--> 

## Enviroment
- **Airbyte version**:  0.27.3-alpha
- **OS Version / Instance**: Ubuntu 16.04.7 LTS
- **Deployment**:  Docker
- **Source Connector and version**: Oracle DB connector: 0.3.1
- **Destination Connector and version**: Snowflake 0.3.11
- **Severity**:  HIGH
- **Step where error happened**:  Setup new connection 

## Current Behavior
Step: Setting source to connect Oracle and Destination to Snowflake, passed checking on source and destination but hangs on setup connection.(there is 1 4G size table on Oracle schema)
 It works on loading csv file on S3 to Snowflake, but cannot work from Oracle -> Snowflake (hangs without error) at set  up connection
it hangs when set up a new connection, checking with Davin (thanks), there is log with  "Caused by: io.grpc.StatusRuntimeException: RESOURCE_EXHAUSTED: grpc: received message larger than max (5420089 vs. 4194304)"
it seem loading too many tables on a db.
however I follow the steps under https://docs.airbyte.io/integrations/sources/oracle#2-create-a-dedicated-read-only-user-with-access-to-the-relevant-tables-recommended-but-optional
created a new user with a table access only, however further checking the log e.g.
2021-07-23 09:15:37 [33mWARN[m i.a.i.s.j.AbstractJdbcSource(lambda$discoverInternal$26):490 - {} - Could not convert column: "START_DATE" from table: "APEX_180200"."APEX_APPLICATION_PAGE_IR_SUB" with type: "-102". Casting to VARCHAR.
It shows up other schema "APEX_180200" (which user cannot be access), so that it may the issue Airbyte scanning all schemas, all tables under a DB while setting new connection, not the tables can be access by the oracle use id.
there are many schema, many tables at the Oracle DB, but I create a new user and 1 table on the same db for extracting tbale using airbyte.


## Expected Behavior
Airbyte should scan or discover the schema the user id can be access/select ( so will not have the issue of "#3943 Discover schema fails when databases have too many tables.", the user id should can access 1 table 1 schema only


## Logs

<details>
<summary> Logs </summary>

```

 "2021-07-23 09:14:49 INFO () DefaultAirbyteStreamFactory(lambda$create$0):73 - 2021-07-23 09:14:49 [32mINFO[m i.a.i.b.IntegrationRunner(run):83 - {} - Integration config: IntegrationConfig{command=DISCOVER, configPath='source_config.json', catalogPath='null', statePath='null'}"


##and  the source_config.json having user id: APP_JEDOX_DATA
{"host":"172.23.10.128","port":51521,"sid":"LFTDWH.SERVER","username":"APP_JEDOX_DATA","password":


##but the log also includes other schema name : APEX_180200
2021-07-23 09:15:37 INFO () DefaultAirbyteStreamFactory(lambda$create$0):73 - 2021-07-23 09:15:37 [33mWARN[m i.a.i.s.j.AbstractJdbcSource(lambda$discoverInternal$26):490 - {} - Could not convert column: "START_DATE" from table: "APEX_180200"."APEX_APPLICATION_PAGE_IR_SUB" with type: "-102". Casting to VARCHAR.
```
</details>

## Steps to Reproduce
1.
2.
3.

## Are you willing to submit a PR?
<!--- 
We accept contributions! 
Don't feel pressured, but if you want to contribute we can help you by giving some tips, 
highlighting the necessary code change or explaining any relevant point your feature will impact.
You can also send questions on #dev Slack channel.

We understand if you can't submit a PR and we're tremendously grateful
that you've already contributed by suggesting a new feature.
-->
Remove this with your answer.

--- a/airbyte-config/init/src/main/resources/seed/source_definitions.yaml
+++ b/airbyte-config/init/src/main/resources/seed/source_definitions.yaml
@@ -289,7 +289,7 @@
 - sourceDefinitionId: b39a7370-74c3-45a6-ac3a-380d48520a83
   name: Oracle DB
   dockerRepository: airbyte/source-oracle
-  dockerImageTag: 0.3.2
+  dockerImageTag: 0.3.3
   documentationUrl: https://docs.airbyte.io/integrations/sources/oracle
 - sourceDefinitionId: c8630570-086d-4a40-99ae-ea5b18673071
   name: Zendesk Talk

--- a/airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcSource.java
+++ b/airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcSource.java
@@ -117,15 +117,13 @@ public abstract class AbstractJdbcSource extends AbstractRelationalDbSource<JDBC
 
   private String getCatalog(SqlDatabase database) {
     return (database.getSourceConfig().has("database") ? database.getSourceConfig().get("database").asText() : null);
-
   }
 
   @Override
-  public List<TableInfo<CommonField<JDBCType>>> discoverInternal(final JdbcDatabase database)
-      throws Exception {
+  protected List<TableInfo<CommonField<JDBCType>>> discoverInternal(JdbcDatabase database, String schema) throws Exception {
     final Set<String> internalSchemas = new HashSet<>(getExcludedInternalNameSpaces());
     return database.bufferedResultSetQuery(
-        conn -> conn.getMetaData().getColumns(getCatalog(database), null, null, null),
+        conn -> conn.getMetaData().getColumns(getCatalog(database), schema, null, null),
         resultSet -> Jsons.jsonNode(ImmutableMap.<String, Object>builder()
             // we always want a namespace, if we cannot get a schema, use db name.
             .put(INTERNAL_SCHEMA_NAME,
@@ -163,6 +161,12 @@ public abstract class AbstractJdbcSource extends AbstractRelationalDbSource<JDBC
                 .collect(Collectors.toList()))
             .build())
         .collect(Collectors.toList());
+  }
+
+  @Override
+  public List<TableInfo<CommonField<JDBCType>>> discoverInternal(final JdbcDatabase database)
+      throws Exception {
+    return discoverInternal(database, null);
   }
 
   @Override

--- a/airbyte-integrations/connectors/source-oracle/Dockerfile
+++ b/airbyte-integrations/connectors/source-oracle/Dockerfile
@@ -9,5 +9,5 @@ COPY build/distributions/${APPLICATION}*.tar ${APPLICATION}.tar
 
 RUN tar xf ${APPLICATION}.tar --strip-components=1
 
-LABEL io.airbyte.version=0.3.2
+LABEL io.airbyte.version=0.3.3
 LABEL io.airbyte.name=airbyte/source-oracle

--- a/airbyte-integrations/connectors/source-oracle/src/main/java/io/airbyte/integrations/source/oracle/OracleSource.java
+++ b/airbyte-integrations/connectors/source-oracle/src/main/java/io/airbyte/integrations/source/oracle/OracleSource.java
@@ -67,6 +67,7 @@ public class OracleSource extends AbstractJdbcSource implements Source {
       configBuilder.put("password", config.get("password").asText());
     }
 
+    // Use the upper-cased username by default.
     schemas = List.of(config.get("username").asText().toUpperCase(Locale.ROOT));
     if (config.has("schemas") && config.get("schemas").isArray()) {
       schemas = new ArrayList<>();
@@ -74,14 +75,12 @@ public class OracleSource extends AbstractJdbcSource implements Source {
         schemas.add(schema.asText());
       }
     }
+
     return Jsons.jsonNode(configBuilder.build());
   }
 
   @Override
   public List<TableInfo<CommonField<JDBCType>>> discoverInternal(JdbcDatabase database) throws Exception {
-    // if schemas is empty
-    // use the user's name
-    // check uniqueness
     List<TableInfo<CommonField<JDBCType>>> internals = new ArrayList<>();
     for (String schema : schemas) {
       LOGGER.debug("Discovering schema: {}", schema);

--- a/airbyte-integrations/connectors/source-oracle/src/main/java/io/airbyte/integrations/source/oracle/OracleSource.java
+++ b/airbyte-integrations/connectors/source-oracle/src/main/java/io/airbyte/integrations/source/oracle/OracleSource.java
@@ -27,10 +27,15 @@ package io.airbyte.integrations.source.oracle;
 import com.fasterxml.jackson.databind.JsonNode;
 import com.google.common.collect.ImmutableMap;
 import io.airbyte.commons.json.Jsons;
+import io.airbyte.db.jdbc.JdbcDatabase;
 import io.airbyte.db.jdbc.OracleJdbcStreamingQueryConfiguration;
 import io.airbyte.integrations.base.IntegrationRunner;
 import io.airbyte.integrations.base.Source;
 import io.airbyte.integrations.source.jdbc.AbstractJdbcSource;
+import io.airbyte.integrations.source.relationaldb.TableInfo;
+import io.airbyte.protocol.models.CommonField;
+import java.sql.JDBCType;
+import java.util.List;
 import java.util.Set;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -59,6 +64,16 @@ public class OracleSource extends AbstractJdbcSource implements Source {
     }
 
     return Jsons.jsonNode(configBuilder.build());
+  }
+
+  @Override
+  public List<TableInfo<CommonField<JDBCType>>> discoverInternal(JdbcDatabase database) throws Exception {
+    // if schemas is empty
+    // use the user's name
+    // check uniqueness
+
+    // otherwise iterate through all the schemas and return them.
+    return super.discoverInternal(database);
   }
 
   @Override

--- a/airbyte-integrations/connectors/source-oracle/src/main/java/io/airbyte/integrations/source/oracle/OracleSource.java
+++ b/airbyte-integrations/connectors/source-oracle/src/main/java/io/airbyte/integrations/source/oracle/OracleSource.java
@@ -35,7 +35,9 @@ import io.airbyte.integrations.source.jdbc.AbstractJdbcSource;
 import io.airbyte.integrations.source.relationaldb.TableInfo;
 import io.airbyte.protocol.models.CommonField;
 import java.sql.JDBCType;
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Locale;
 import java.util.Set;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -45,6 +47,8 @@ public class OracleSource extends AbstractJdbcSource implements Source {
   private static final Logger LOGGER = LoggerFactory.getLogger(OracleSource.class);
 
   static final String DRIVER_CLASS = "oracle.jdbc.OracleDriver";
+
+  private List<String> schemas;
 
   public OracleSource() {
     super(DRIVER_CLASS, new OracleJdbcStreamingQueryConfiguration());
@@ -63,6 +67,13 @@ public class OracleSource extends AbstractJdbcSource implements Source {
       configBuilder.put("password", config.get("password").asText());
     }
 
+    schemas = List.of(config.get("username").asText().toUpperCase(Locale.ROOT));
+    if (config.has("schemas") && config.get("schemas").isArray()) {
+      schemas = new ArrayList<>();
+      for (final JsonNode schema : config.get("schemas")) {
+        schemas.add(schema.asText());
+      }
+    }
     return Jsons.jsonNode(configBuilder.build());
   }
 
@@ -71,9 +82,17 @@ public class OracleSource extends AbstractJdbcSource implements Source {
     // if schemas is empty
     // use the user's name
     // check uniqueness
+    List<TableInfo<CommonField<JDBCType>>> internals = new ArrayList<>();
+    for (String schema : schemas) {
+      LOGGER.debug("Discovering schema: {}", schema);
+      internals.addAll(super.discoverInternal(database, schema));
+    }
 
-    // otherwise iterate through all the schemas and return them.
-    return super.discoverInternal(database);
+    for (TableInfo<CommonField<JDBCType>> info : internals) {
+      LOGGER.debug("Found table: {}", info.getName());
+    }
+
+    return internals;
   }
 
   @Override

--- a/airbyte-integrations/connectors/source-oracle/src/main/resources/spec.json
+++ b/airbyte-integrations/connectors/source-oracle/src/main/resources/spec.json
@@ -35,6 +35,16 @@
         "description": "Password associated with the username.",
         "type": "string",
         "airbyte_secret": true
+      },
+      "schemas": {
+        "title": "Schemas",
+        "description": "List of schemas to sync from. Defaults to user.",
+        "type": "array",
+        "items": {
+          "type": "string"
+        },
+        "minItems": 1,
+        "uniqueItems": true
       }
     }
   }

--- a/airbyte-integrations/connectors/source-oracle/src/main/resources/spec.json
+++ b/airbyte-integrations/connectors/source-oracle/src/main/resources/spec.json
@@ -38,7 +38,7 @@
       },
       "schemas": {
         "title": "Schemas",
-        "description": "List of schemas to sync from. Defaults to user.",
+        "description": "List of schemas to sync from. Defaults to user. Case sensitive.",
         "type": "array",
         "items": {
           "type": "string"

--- a/airbyte-integrations/connectors/source-oracle/src/test-integration/java/io/airbyte/integrations/source/oracle/OracleSourceAcceptanceTest.java
+++ b/airbyte-integrations/connectors/source-oracle/src/test-integration/java/io/airbyte/integrations/source/oracle/OracleSourceAcceptanceTest.java
@@ -64,6 +64,7 @@ public class OracleSourceAcceptanceTest extends SourceAcceptanceTest {
         .put("sid", container.getSid())
         .put("username", container.getUsername())
         .put("password", container.getPassword())
+        .put("schemas", List.of("JDBC_SPACE"))
         .build());
 
     JdbcDatabase database = Databases.createJdbcDatabase(config.get("username").asText(),

--- a/airbyte-integrations/connectors/source-oracle/src/test-integration/java/io/airbyte/integrations/source/oracle/OracleSourceDatatypeTest.java
+++ b/airbyte-integrations/connectors/source-oracle/src/test-integration/java/io/airbyte/integrations/source/oracle/OracleSourceDatatypeTest.java
@@ -39,6 +39,7 @@ import java.text.SimpleDateFormat;
 import java.time.Instant;
 import java.util.Calendar;
 import java.util.Date;
+import java.util.List;
 import java.util.TimeZone;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -62,6 +63,7 @@ public class OracleSourceDatatypeTest extends AbstractSourceDatabaseTypeTest {
         .put("sid", container.getSid())
         .put("username", container.getUsername())
         .put("password", container.getPassword())
+        .put("schemas", List.of("TEST"))
         .build());
 
     Database database = Databases.createOracleDatabase(config.get("username").asText(),

--- a/airbyte-integrations/connectors/source-oracle/src/test/java/io/airbyte/integrations/source/oracle/OracleJdbcSourceAcceptanceTest.java
+++ b/airbyte-integrations/connectors/source-oracle/src/test/java/io/airbyte/integrations/source/oracle/OracleJdbcSourceAcceptanceTest.java
@@ -37,6 +37,7 @@ import java.sql.DriverManager;
 import java.sql.ResultSet;
 import java.sql.SQLException;
 import java.sql.Statement;
+import java.util.List;
 import org.junit.jupiter.api.AfterAll;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeAll;
@@ -88,6 +89,7 @@ class OracleJdbcSourceAcceptanceTest extends JdbcSourceAcceptanceTest {
         .put("sid", ORACLE_DB.getSid())
         .put("username", ORACLE_DB.getUsername())
         .put("password", ORACLE_DB.getPassword())
+        .put("schemas", List.of(SCHEMA_NAME, SCHEMA_NAME2))
         .build());
 
     // Because Oracle doesn't let me create database easily I need to clean up

--- a/airbyte-integrations/connectors/source-oracle/src/test/java/io/airbyte/integrations/source/oracle/OracleSourceTest.java
+++ b/airbyte-integrations/connectors/source-oracle/src/test/java/io/airbyte/integrations/source/oracle/OracleSourceTest.java
@@ -70,7 +70,6 @@ class OracleSourceTest {
 
   private static OracleContainer ORACLE_DB;
 
-  private static OracleContainer container;
   private static JsonNode config;
 
   @BeforeAll
@@ -87,6 +86,7 @@ class OracleSourceTest {
         .put("sid", ORACLE_DB.getSid())
         .put("username", ORACLE_DB.getUsername())
         .put("password", ORACLE_DB.getPassword())
+        .put("schemas", List.of("TEST"))
         .build());
 
     JdbcDatabase database = Databases.createJdbcDatabase(config.get("username").asText(),
@@ -105,16 +105,6 @@ class OracleSourceTest {
     });
 
     database.close();
-  }
-
-  private JdbcDatabase getDatabaseFromConfig(JsonNode config) {
-    return Databases.createJdbcDatabase(config.get("username").asText(),
-        config.get("password").asText(),
-        String.format("jdbc:oracle:thin:@//%s:%s/%s",
-            config.get("host").asText(),
-            config.get("port").asText(),
-            config.get("sid").asText()),
-        "oracle.jdbc.driver.OracleDriver");
   }
 
   private JsonNode getConfig(OracleContainer oracleDb) {
@@ -142,7 +132,7 @@ class OracleSourceTest {
 
   @Test
   void testReadSuccess() throws Exception {
-    final Set<AirbyteMessage> actualMessages = MoreIterators.toSet(new OracleSource().read(getConfig(ORACLE_DB), CONFIGURED_CATALOG, null));
+    final Set<AirbyteMessage> actualMessages = MoreIterators.toSet(new OracleSource().read(config, CONFIGURED_CATALOG, null));
     setEmittedAtToNull(actualMessages);
 
     assertEquals(ASCII_MESSAGES, actualMessages);

--- a/airbyte-integrations/connectors/source-relational-db/src/main/java/io/airbyte/integrations/source/relationaldb/AbstractRelationalDbSource.java
+++ b/airbyte-integrations/connectors/source-relational-db/src/main/java/io/airbyte/integrations/source/relationaldb/AbstractRelationalDbSource.java
@@ -132,6 +132,17 @@ public abstract class AbstractRelationalDbSource<DataType, Database extends SqlD
       throws Exception;
 
   /**
+   * Discovers all available tables within a schema in the source database.
+   *
+   * @param database - source database
+   * @param schema - source schema
+   * @return list of source tables
+   * @throws Exception - access to the database might lead to exceptions.
+   */
+  protected abstract List<TableInfo<CommonField<DataType>>> discoverInternal(final Database database, String schema)
+      throws Exception;
+
+  /**
    * Discover Primary keys for each table and @return a map of namespace.table name to their
    * associated list of primary key fields.
    *

--- a/docs/integrations/sources/oracle.md
+++ b/docs/integrations/sources/oracle.md
@@ -96,9 +96,14 @@ GRANT SELECT ON "<schema_b>"."<table_2>" TO airbyte;
 
 Your database user should now be ready for use with Airbyte.
 
+#### 3. Include the schemas Airbyte should look at when configuring the Airbyte Oracle Source.
+
+Case sensitive. Defaults to the upper-cased user if empty. If the user does not have access to the configured schemas, no tables will be discovered.
+
 
 ## Changelog
 
 | Version | Date       | Pull Request | Subject |
 | :------ | :--------  | :-----       | :------ |
-| 0.3.2   | 2021-08-13 | [4699](https://github.com/airbytehq/airbyte/pull/4699) | Added json config validator |
+| 0.3.3   | 2021-09-01 | [5779](https://github.com/airbytehq/airbyte/pull/5779) | Ability to only discover certain schemas. |
+| 0.3.2   | 2021-08-13 | [4699](https://github.com/airbytehq/airbyte/pull/4699) | Added json config validator. |

